### PR TITLE
Add JSON error handling

### LINKED ISSUE #146: Ensure all supported WordPress characters are translated correctly to Apple News
See: https://wordpress.org/support/topic/character-encoding-4

### PR DESCRIPTION
**Pull request by bcampeau**
1. Creates a framework for JSON validation and initially introduces checking for unicode character sequences that are symptomatic of display issues that have been witnessed, though not reproduced, in Apple News.
2. Breaks out Action_Exception into its own class file for cleanliness.
3. Adds direct support links to every error message.
4. Adds better formatting of multiple error messages.
5. Adds unit tests for the Apple_News and Admin_Apple_Notice classes
6. Adds new unit tests for Push

Closes #146

### CODE REVIEW

File: includes/class-apple-news.php [hunk 1]
```diff
@@ -86,4 +102,34 @@ public function migrate_settings( $settings ) {
...
+	 * @var string $format
+	 * @var boolean $with_padding
+	 * @return string
+	 * @access public
+	 */
+	public static function get_support_info( $format = 'html', $with_padding = true ) {
+		$support_info = sprintf(
+			__( 'If you need assistance with this issue, please reach out for support on <a href="%s">WordPress.org</a> or <a href="%s">github</a>.', 'apple-news' ),
+			esc_attr( self::$wordpress_org_support_url ),
+			esc_attr( self::$github_support_url )
```
Comment (mslinnea) [inline includes/class-apple-news.php]: I think these 2 lines should actually be `esc_url()`, although the values are hardcoded, so not sure if any escaping is really necessary

File: tests/admin/apple-actions/index/test-class-push.php [hunk 2]
```diff
@@ -258,22 +255,103 @@ public function testComponentErrorsFail() {
...
+			$this->assertEquals( null, get_post_meta( $post_id, 'apple_news_api_deleted', true ) );
+		}
+	}
+
+	public function testJSONErrorsWarn() {
+		$this->settings->set( 'component_alerts', 'none' );
+		$this->settings->set( 'json_alerts', 'warn' );
+
+		$response = $this->dummy_response();
+		$api = $this->prophet->prophesize( '\Apple_Push_API\API' );
```
Comment (mslinnea) [inline tests/admin/apple-actions/index/test-class-push.php]: ooh cool name for a function

File: tests/test-class-apple-news.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		$this->assertEquals( 'image.jpg', $filename );
+	}
+
+	public function testVersion() {
+		$plugin_data = apple_news_get_plugin_data();
+		$this->assertEquals( $plugin_data['Version'], Apple_News::$version );
+	}
+
+	public function testMigrateSettings() {
+		update_option( 'use_remote_images', 'yes' );
```
Comment (mslinnea) [inline tests/test-class-apple-news.php]: Do you need to revert this option back to its original value after running this test?  Maybe this is handled somewhere else, or maybe it's not necessary, but just thought I'd ask.

**Comment by mslinnea** [comment]
🐊

File: includes/class-apple-news.php (hunk 1 above)
Comment (bcampeau) [inline includes/class-apple-news.php]: Good point

Comment (bcampeau) [inline includes/class-apple-news.php]: It's always best to escape variables, even if static. If they become non-static later, then the escaping would already be in place.

File: tests/admin/apple-actions/index/test-class-push.php (hunk 2 above)
Comment (bcampeau) [inline tests/admin/apple-actions/index/test-class-push.php]: It's part of a unit test library called Prophecy. It allows you to instantiate dummy versions of complex classes, initialize with sample data when exact values aren't relevant, check simply whether a function executes but not necessarily execute with any specific value. It's very useful.

File: tests/test-class-apple-news.php (hunk 3 above)
Comment (bcampeau) [inline tests/test-class-apple-news.php]: No. A new instance of WP is spun up for all unit tests and is destroyed at the end. Values aren't persistent.

**Pull request merged**